airbyte-integrations/bases/base-typing-deduping/src/main/java/io/airbyte/integrations/base/destination/typing_deduping/AirbyteTypeUtils.java

```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.integrations.base.destination.typing_deduping;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.node.TextNode;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import io.airbyte.integrations.base.destination.typing_deduping.AirbyteType.AirbyteProtocolType;
+import io.airbyte.integrations.base.destination.typing_deduping.AirbyteType.Array;
+import io.airbyte.integrations.base.destination.typing_deduping.AirbyteType.OneOf;
+import io.airbyte.integrations.base.destination.typing_deduping.AirbyteType.Struct;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Stream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class AirbyteTypeUtils {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(AirbyteTypeUtils.class);
+
+  // Map from a protocol type to what other protocol types should take precedence over it if present
+  // in a OneOf
+  private static final Map<AirbyteProtocolType, List<AirbyteProtocolType>> EXCLUDED_PROTOCOL_TYPES_MAP = ImmutableMap.of(
+      AirbyteProtocolType.BOOLEAN, ImmutableList.of(AirbyteProtocolType.STRING, AirbyteProtocolType.NUMBER, AirbyteProtocolType.INTEGER),
+      AirbyteProtocolType.INTEGER, ImmutableList.of(AirbyteProtocolType.STRING, AirbyteProtocolType.NUMBER),
+      AirbyteProtocolType.NUMBER, ImmutableList.of(AirbyteProtocolType.STRING));
+
+  // Protocol types in order of precedence
+  private static final List<AirbyteProtocolType> ORDERED_PROTOCOL_TYPES = ImmutableList.of(
+      AirbyteProtocolType.BOOLEAN,
+      AirbyteProtocolType.INTEGER,
+      AirbyteProtocolType.NUMBER,
+      AirbyteProtocolType.TIMESTAMP_WITHOUT_TIMEZONE,
+      AirbyteProtocolType.TIMESTAMP_WITH_TIMEZONE,
+      AirbyteProtocolType.DATE,
+      AirbyteProtocolType.TIME_WITH_TIMEZONE,
+      AirbyteProtocolType.TIME_WITHOUT_TIMEZONE,
+      AirbyteProtocolType.STRING);
+
+  protected static boolean nodeIsType(final JsonNode node, final String type) {
+    if (node == null || !node.isTextual()) {
+      return false;
+    }
+    return node.equals(TextNode.valueOf(type));
+  }
+
+  private static boolean nodeIsOrContainsType(final JsonNode node, final String type) {
+    if (node == null) {
+      return false;
+    } else if (node.isTextual()) {
+      return nodeIsType(node, type);
+    } else if (node.isArray()) {
+      for (final JsonNode element : node) {
+        if (nodeIsType(element, type)) {
+          return true;
+        }
+      }
+    }
+    return false;
+  }
+
+  protected static AirbyteType getAirbyteProtocolType(final JsonNode node) {
```

Comment:
Is it safe to assume this method is tested by the `AirbyteTypeTest.java` tests?

Reply:
Yes this is spot-tested here: https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/bases/base-typing-deduping/src/test/java/io/airbyte/integrations/base/destination/typing_deduping/AirbyteTypeTest.java#L426-L440